### ISSUE TITLE
How can I use dino pre-trained model with video clips as mentioned in the blog post?

### ISSUE THREAD

**Issue by shamanez** [facebookresearch/dino#2]

**Comment by loretoparisi** [comment]
Same question here: how to run inference on a sample image. I was looking at eval script, but I assume that a simple segmentation inference it's easier.

**Comment by aquadzn** [comment]
https://user-images.githubusercontent.com/46140458/116817046-f0cd5580-ab64-11eb-8b56-18c811f7883e.mp4

this is what I can get using ffmpeg
I can make a notebook if you're interested

**Comment by maxrohleder** [comment]
Hey @aquadzn, a notebook would be cool! Is it just me or does the left video freeze after ~6 seconds?

**Comment by aquadzn** [comment]
yes because there were 6 videos but it's good now, I blended them into one:

https://user-images.githubusercontent.com/46140458/116817761-47885e80-ab68-11eb-9975-d61d5a919e13.mp4

**Comment by aquadzn** [comment]
@loretoparisi @shamanez @maxrohleder you can try with this
https://gist.github.com/aquadzn/32ac53aa6e485e7c3e09b1a0914f7422

**Comment by mathildecaron31** [comment]
Really cool, thanks a lot @aquadzn !!

Also feel free to propose a PR to extend `visualize_attention.py` which supports visualization for an image only atm.

@loretoparisi 
>how to run inference on a sample image
You can run `python visualize_attention.py`
https://github.com/facebookresearch/dino#self-attention-visualization

**Comment by aquadzn** [comment]
> Really cool, thanks a lot @aquadzn !!
> 
> Also feel free to propose a PR to extend `visualize_attention.py` which supports visualization for an image only atm.
> 
> @loretoparisi
> 
> > how to run inference on a sample image
> 
> You can run `python visualize_attention.py`
> https://github.com/facebookresearch/dino#self-attention-visualization

https://github.com/facebookresearch/dino/pull/12

**Comment by cishwarya** [comment]
Dear @aquadzn I have problem running the inference script, https://gist.github.com/aquadzn/32ac53aa6e485e7c3e09b1a0914f7422

The images I use are orginally grayscale and have been converted to 3 channels. Now the shape of my images are (427, 488, 3)

When I run **`predict_video(args)`** I get the following error. Could you please help me here? thanks


---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-108-60e00ae57b11> in <module>()
----> 1 predict_video(args)

6 frames
<ipython-input-104-91499e0fd0de> in predict_video(args)
     41 
     42 
---> 43         attentions = model.forward_selfattention(img.cuda())
     44 
     45         nh = attentions.shape[1] # number of head

~/vision_transformer.py in forward_selfattention(self, x)
    220         B, nc, w, h = x.shape
    221         N = self.pos_embed.shape[1] - 1
--> 222         x = self.patch_embed(x)
    223 
    224         # interpolate patch embeddings

/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    887             result = self._slow_forward(*input, **kwargs)
    888         else:
--> 889             result = self.forward(*input, **kwargs)
    890         for hook in itertools.chain(
    891                 _global_forward_hooks.values(),

~/vision_transformer.py in forward(self, x)
    116     def forward(self, x):
    117         B, C, H, W = x.shape
--> 118         x = self.proj(x).flatten(2).transpose(1, 2)
    119         return x
    120 

/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    887             result = self._slow_forward(*input, **kwargs)
    888         else:
--> 889             result = self.forward(*input, **kwargs)
    890         for hook in itertools.chain(
    891                 _global_forward_hooks.values(),

/usr/local/lib/python3.7/dist-packages/torch/nn/modules/conv.py in forward(self, input)
    397 
    398     def forward(self, input: Tensor) -> Tensor:
--> 399         return self._conv_forward(input, self.weight, self.bias)
    400 
    401 class Conv3d(_ConvNd):

/usr/local/lib/python3.7/dist-packages/torch/nn/modules/conv.py in _conv_forward(self, input, weight, bias)
    394                             _pair(0), self.dilation, self.groups)
    395         return F.conv2d(input, weight, bias, self.stride,
--> 396                         self.padding, self.dilation, self.groups)
    397 
    398     def forward(self, input: Tensor) -> Tensor:

RuntimeError: Given groups=1, weight of size [384, 3, 8, 8], expected input[1, 0, 512, 585] to have 3 channels, but got 0 channels instead

**Comment by aquadzn** [comment]
@cishwarya are you sure that your image is correctly converted to 3 channels? because it says input[1, 0, 512, 585]

**Comment by cishwarya** [comment]
Hi @aquadzn thanks for your reply. As you could see from the below code, inside **predict_video(args)** function, I tried to print the shape of the image and it shows [3, 512, 585].

```
import numpy as np
import cv2
import glob
def predict_video(args):

    for filepath in sorted(glob.iglob(args.image_path + "/*")):
        print("filepath",filepath)
        img = Image.open(filepath)
        img = img.convert('RGB')
        transform = pth_transforms.Compose([
            pth_transforms.ToTensor(),
            pth_transforms.Resize(512),
            pth_transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
        ])
        img = transform(img)

        print(".img.shape",img.shape)  # output is torch.Size([3, 512, 585])
```

**Comment by himanshu-daga** [comment]
> yes because there were 6 videos but it's good now, I blended them into one:
> 
>  example.mp4

Hi @aquadzn,
I'm currently getting the o/p video similar to 6sec to 12sec in your blended video. How can one move from that to the o/p you get in 0sec to 6sec?

**Comment by aquadzn** [comment]
> > yes because there were 6 videos but it's good now, I blended them into one:
> > example.mp4
> 
> Hi @aquadzn,
> I'm currently getting the o/p video similar to 6sec to 12sec in your blended video. How can one move from that to the o/p you get in 0sec to 6sec? I did try different thresholds but didn't get similar results..

Hello, I don't understand, you mean that you got a video like the first one I shared instead of the second one? Did you use my script `video_generation.py` ?

**Comment by himanshu-daga** [comment]
I used the colab version (https://gist.github.com/aquadzn/32ac53aa6e485e7c3e09b1a0914f7422) to generate the video. PFB my resulting video. I want to get a mask only for the chair but as you can see in the video, it's detecting many more objects. How can I extract a mask only for the chair?

https://user-images.githubusercontent.com/25201878/117499164-76536a00-af98-11eb-9718-3e9bb647e83e.mp4

**Comment by hobbitsyfeet** [comment]
I was able to generate a video of my own with `video_generation.py` as mentioned in https://github.com/facebookresearch/dino/issues/2#issuecomment-834584669  and the instructions provided. I don't have enough V-RAM so I needed to either:

1. Use a `--patch_size of 16` rather than 8. This requires you to download `dino_vitbase16_pretrain_full_checkpoint.pth` and use `--arch vit_base` `--pretrained_weights dino_vitbase16_pretrain_full_checkpoint.pth`
2. `--resize` the video to a lower resolution, I did `--resize 480 720` (720 width, 480 height)

An example:
`python video_generation.py --pretrained_weights ./dino_vitbase8_pretrain_full_checkpoint.pth --input_path ./video.mp4 --output_path ./output --patch_size 8 --arch vit_base --resize 480 720`

I used Anaconda to install PyTorch and the requirements. I have CUDA 10.1 installed with the appropriate CUDNN (v8.05).
[PyTorch site](https://pytorch.org/get-started/locally/) shows you suggested installs, mine was 
`conda install pytorch torchvision torchaudio cudatoolkit=10.2 -c pytorch` (it still worked even though I have CUDA 10.1, hmm)

And extra requirements for the project
```
conda install -c conda-forge opencv
conda install -c conda-forge tqdm
conda install -c conda-forge matplotlib
```

Don't know if this is the issue you all mention, but it might be useful to someone.
